### ISSUE TITLE
Problem with --delrate since postfwd 2.0

### ISSUE THREAD

**Issue by ArnGa** [postfwd/postfwd#26]
Hello,

Since Postfwd 2.0, the --delrate does not work anymore.

###################################################
With Postfwd 1.39:
```
# postfwd --dumpcache

<undef>

```
=> connection from one client

```
# postfwd --dumpcache

%rate_cache -> %client_address=192.168.100.155 -> @list                -> 'CLASS301+1000_86400,CLASS302+500_86400'
%rate_cache -> %client_address=192.168.100.155 -> %CLASS301+1000_86400 -> @action   -> 'REJECT limit of rcpts per day for client 192.168.100.155 reached: CLASS 3'
%rate_cache -> %client_address=192.168.100.155 -> %CLASS301+1000_86400 -> @count    -> '1'
%rate_cache -> %client_address=192.168.100.155 -> %CLASS301+1000_86400 -> @maxcount -> '1000'
%rate_cache -> %client_address=192.168.100.155 -> %CLASS301+1000_86400 -> @rule     -> 'CLASS301'
%rate_cache -> %client_address=192.168.100.155 -> %CLASS301+1000_86400 -> @time     -> '1561390660.49497'
%rate_cache -> %client_address=192.168.100.155 -> %CLASS301+1000_86400 -> @ttl      -> '86400'
%rate_cache -> %client_address=192.168.100.155 -> %CLASS301+1000_86400 -> @type     -> 'rcpt'
%rate_cache -> %client_address=192.168.100.155 -> %CLASS301+1000_86400 -> @until    -> '1561477060.49497'
%rate_cache -> %client_address=192.168.100.155 -> %CLASS302+500_86400  -> @action   -> 'WARN state WARN for client 192.168.100.155: CLASS 3'
%rate_cache -> %client_address=192.168.100.155 -> %CLASS302+500_86400  -> @count    -> '1'
%rate_cache -> %client_address=192.168.100.155 -> %CLASS302+500_86400  -> @maxcount -> '500'
%rate_cache -> %client_address=192.168.100.155 -> %CLASS302+500_86400  -> @rule     -> 'CLASS302'
%rate_cache -> %client_address=192.168.100.155 -> %CLASS302+500_86400  -> @time     -> '1561390660.49497'
%rate_cache -> %client_address=192.168.100.155 -> %CLASS302+500_86400  -> @ttl      -> '86400'
%rate_cache -> %client_address=192.168.100.155 -> %CLASS302+500_86400  -> @type     -> 'rcpt'
%rate_cache -> %client_address=192.168.100.155 -> %CLASS302+500_86400  -> @until    -> '1561477060.49497'
```

```
# postfwd --delrate client_address=192.168.100.155

rate cache item 'client_address=192.168.100.155' removed
```

=> OK

########################################################
Same operation with Postfwd 2.0 (I tested all versions 2.0 until the last 2.02):
```
# postfwd --dumpcache

<undef>

```
=> connection from one client

```
# postfwd --dumpcache

%Request_Cache -> %NRrL9YgKuBcCVJhVfxKMvA -> $action -> 'OK'
%Request_Cache -> %NRrL9YgKuBcCVJhVfxKMvA -> $hit    -> '2'
%Request_Cache -> %NRrL9YgKuBcCVJhVfxKMvA -> $hits   -> 'CLASS300;CLASS399'
%Request_Cache -> %NRrL9YgKuBcCVJhVfxKMvA -> $id     -> 'CLASS399'
%Request_Cache -> %NRrL9YgKuBcCVJhVfxKMvA -> $ttl    -> '600'
%Request_Cache -> %NRrL9YgKuBcCVJhVfxKMvA -> $until  -> '1561384037.55964'
%Request_Cache -> %fEzkLXt5hRGQGhIhH2furg -> $action -> 'OK'
%Request_Cache -> %fEzkLXt5hRGQGhIhH2furg -> $hit    -> '2'
%Request_Cache -> %fEzkLXt5hRGQGhIhH2furg -> $hits   -> 'CLASS300;CLASS301;CLASS302;CLASS399'
%Request_Cache -> %fEzkLXt5hRGQGhIhH2furg -> $id     -> 'CLASS399'
%Request_Cache -> %fEzkLXt5hRGQGhIhH2furg -> $ttl    -> '600'
%Request_Cache -> %fEzkLXt5hRGQGhIhH2furg -> $until  -> '1561384039.94785'
%Rate_Cache -> %client_address=192.168.100.155 -> @list                -> 'CLASS301+1000_86400,CLASS302+500_86400'
%Rate_Cache -> %client_address=192.168.100.155 -> %CLASS301+1000_86400 -> $action   -> 'REJECT limit of rcpts per day for client 192.168.100.155 reached: CLASS 3'
%Rate_Cache -> %client_address=192.168.100.155 -> %CLASS301+1000_86400 -> $count    -> '1'
%Rate_Cache -> %client_address=192.168.100.155 -> %CLASS301+1000_86400 -> $maxcount -> '1000'
%Rate_Cache -> %client_address=192.168.100.155 -> %CLASS301+1000_86400 -> $rule     -> 'CLASS301'
%Rate_Cache -> %client_address=192.168.100.155 -> %CLASS301+1000_86400 -> $time     -> '1561383439.94785'
%Rate_Cache -> %client_address=192.168.100.155 -> %CLASS301+1000_86400 -> $ttl      -> '86400'
%Rate_Cache -> %client_address=192.168.100.155 -> %CLASS301+1000_86400 -> $type     -> 'rcpt'
%Rate_Cache -> %client_address=192.168.100.155 -> %CLASS301+1000_86400 -> $until    -> '1561469839.94785'
%Rate_Cache -> %client_address=192.168.100.155 -> %CLASS302+500_86400  -> $action   -> 'WARN state WARN for client 192.168.100.155: CLASS 3'
%Rate_Cache -> %client_address=192.168.100.155 -> %CLASS302+500_86400  -> $count    -> '1'
%Rate_Cache -> %client_address=192.168.100.155 -> %CLASS302+500_86400  -> $maxcount -> '500'
%Rate_Cache -> %client_address=192.168.100.155 -> %CLASS302+500_86400  -> $rule     -> 'CLASS302'
%Rate_Cache -> %client_address=192.168.100.155 -> %CLASS302+500_86400  -> $time     -> '1561383439.94785'
%Rate_Cache -> %client_address=192.168.100.155 -> %CLASS302+500_86400  -> $ttl      -> '86400'
%Rate_Cache -> %client_address=192.168.100.155 -> %CLASS302+500_86400  -> $type     -> 'rcpt'
%Rate_Cache -> %client_address=192.168.100.155 -> %CLASS302+500_86400  -> $until    -> '1561469839.94785'
```

```
# postfwd --delrate client_address=192.168.100.155

rate cache removal of 'client_address=192.168.100.155' failed: item not found
```

=> NOK

The entry is not deleted.

I feel that this is a bug and I therefore remain in 1.39 for now.

Regards,

**Comment by FredNass** [comment]
Hello Jan,

If you happen to be here... I can confirm that v2.02 is not working as expected on my RHEL 8.
And I think this particular issue that ArnGa reported has something to do with it. With v2.02, rate cache entries are not automatically cleared on a time basis anymore and once a rule has matched, the only way it's cleared is by waiting for --cache time to be over.

Frédéric.

**Comment by jpkessler** [comment]
Hey guys, sorry for that late reply. Fixed with v2.03.
